DBZ-7300 Review PostgreSQL documentations with new Snapshotter interfaces (#5213)
debezium/debezium: documentation/modules/ROOT/pages/connectors/postgresql.adoc

     */
    void validate(boolean offsetContextExists, boolean isSnapshotInProgress);

    /**
     * @return {@code true} if the snapshotter should take a snapshot
     */
    boolean shouldSnapshot();

    /**
     * @return {@code true} if the snapshotter should take a snapshot
     */
    boolean shouldSnapshotSchema();

    /**
     * @return {@code true} if the snapshotter should stream after taking a snapshot
     */
    boolean shouldStream();

    /**
     * @return {@code true} whether the schema can be recovered if database schema history is corrupted.
     */
    boolean shouldSnapshotOnSchemaError();

    /**
     * @return {@code true} whether the snapshot should be re-executed when there is a gap in data stream.
     */
    boolean shouldSnapshotOnDataError();

    /**
     *
     * @return {@code true} if streaming should resume from the start of the snapshot
     * transaction, or {@code false} for when a connector resumes and takes a snapshot,
     * streaming should resume from where streaming previously left off.
     */
    default boolean shouldStreamEventsStartingFromSnapshot() {
        return true;
    }

    /**

>      * Lifecycle hook called once the snapshot phase is successful.

Reviewer: ```suggestion
     * Lifecycle hook called after the snapshot phase is successful.
```

     */
    default void snapshotCompleted() {
        // no operation
    }

    /**

>      * Lifecycle hook called once the snapshot phase is aborted.

Reviewer: ```suggestion
     * Lifecycle hook called after the snapshot phase is aborted.
```

     */
    default void snapshotAborted() {
        // no operation
    }
}
----

[source,java,indent=0,subs="+attributes"]
.io.debezium.snapshot.spi.SnapshotQuery interface. All built-in snapshot query modes implement this interface.
----
/**

>  * {@link SnapshotQuery} is used to determine the query used during data snapshot

Reviewer: ```suggestion
 * {@link SnapshotQuery} is used to determine the query used during a data snapshot
```

`exported` -  deprecated +
 +
ifdef::community[]

> `custom` - The connector performs a snapshot according to the implementation provided via `snapshot.mode.custom.name` property, which is a custom implementation of the `io.debezium.spi.snapshot.Snapshotter` interface. +

Reviewer: ```suggestion
`custom` - The connector performs a snapshot according to the implementation specified by the xref:postgresql-property-snapshot-mode-custom-name[`snapshot.mode.custom.name`] property, which defines a custom implementation of the `io.debezium.spi.snapshot.Snapshotter` interface. +
```

endif::community[]
 +
For more information, see the xref:#postgresql-connector-snapshot-mode-options[table of `snapshot.mode` options].

ifdef::community[]
|[[postgresql-property-snapshot-mode-custom-name]]<<postgresql-property-snapshot-mode-custom-name, `+snapshot.mode.custom.name+`>>
|No default

> | When `snapshot.mode` is set as `custom`, this setting must be set to specify the name of the custom implementation provided in the `name()` method defined by the 'io.debezium.spi.snapshot.Snapshotter' interface.

Reviewer: ```suggestion
| When `snapshot.mode` is set as `custom`, use this setting to specify the name of the custom implementation provided in the `name()` method that is defined by the 'io.debezium.spi.snapshot.Snapshotter' interface.
```

> The provided implementation is called on each app boot to determine whether to do a snapshot. See xref:postgresql-custom-snapshot[custom snapshotter SPI].

Reviewer: ```suggestion
The provided implementation is called after a connector restart to determine whether to perform a snapshot. 
For more information, see xref:postgresql-custom-snapshot[custom snapshotter SPI].
```

endif::community[]

|[[postgresql-property-snapshot-locking-mode]]<<postgresql-property-snapshot-locking-mode, `+snapshot.locking.mode+`>>
|`none`

> |Controls how the connector holds locks on tables while performing the schema snapshot: +

Reviewer: ```suggestion
|Specifies how the connector holds locks on tables while performing a schema snapshot: +
Set one of the following options:
```

 +

> `shared` -  The connector will hold a table lock that prevents exclusive table access for just the initial portion of the snapshot while the database schemas and other metadata are being read. The remaining work in a snapshot involves selecting all rows from each table, and this is done using a flashback query that requires no locks. +
>  +
> `none` -  The connector avoids locks entirely. This mode is only safe to use if no schema changes are happening while the snapshot is taken. +
>  +
> ifdef::community[]
> `custom` - The connector performs a snapshot according to the implementation provided via `snapshot.locking.mode.custom.name` property, which is a custom implementation of the `io.debezium.spi.snapshot.SnapshotLock` interface. +
> endif::community[]

Reviewer: ```suggestion
`shared`:: The connector holds a table lock that prevents exclusive table access during the initial portion phase of the snapshot in which database schemas and other metadata are read. 
After the initial phase, the snapshot no longer requires table locks. 
 +
`none`:: The connector avoids locks entirely. +
+
[WARNING]
====
Do not use this mode if schema changes might occur during the snapshot. 
====
 
ifdef::community[]
`custom`:: The connector performs a snapshot according to the implementation specified by the xref:postgresql-property-snapshot-locking-mode-custom-name[`snapshot.locking.mode.custom.name`] property, which is a custom implementation of the `io.debezium.spi.snapshot.SnapshotLock` interface.
endif::community[]
```

ifdef::community[]
|[[postgresql-property-snapshot-locking-mode-custom-name]]<<postgresql-property-snapshot-locking-mode-custom-name, `+snapshot.locking.mode.custom.name+`>>
|No default

> | When `snapshot.locking.mode` is set as `custom`, this setting must be set to specify the name of the custom implementation provided in the `name()` method defined by the 'io.debezium.spi.snapshot.SnapshotLock' interface.

Reviewer: ```suggestion
| When `snapshot.locking.mode` is set as `custom`, use this setting to specify the name of the custom implementation provided in the `name()` method that is defined by the 'io.debezium.spi.snapshot.SnapshotLock' interface.
```

> See xref:postgresql-custom-snapshot[custom snapshotter SPI].

Reviewer: ```suggestion
For more information, see xref:postgresql-custom-snapshot[custom snapshotter SPI].
```

endif::community[]

|[[postgresql-property-snapshot-query-mode]]<<postgresql-property-snapshot-query-mode, `+snapshot.query.mode+`>>

> |`select_all`
> |Controls how the connector queries data while performing the snapshot: +
>  +
> 
> `select_all` - The connector performs a select all query. +
>  +
> ifdef::community[]
> `custom` - The connector performs a snapshot query according to the implementation provided via `snapshot.query.mode.custom.name` property, which is a custom implementation of the `io.debezium.spi.snapshot.SnapshotQuery` interface. +
> endif::community[]
>  +
> This is a more flexible alternative to xref:{context}-property-snapshot-select-statement-overrides[`snapshot.select.statement.overrides`]

Reviewer: ```suggestion
|`select_all`
|Specifies how the connector queries data while performing a snapshot. +
Set one of the following options:

`select_all`:: The connector performs a `select all` query. 

ifdef::community[]
`custom`:: The connector performs a snapshot query according to the implementation specified by the xref:postgresql-property-snapshot-snapshot-query-mode-custom-name[`snapshot.query.mode.custom.name`] property, which defines a custom implementation of the `io.debezium.spi.snapshot.SnapshotQuery` interface. +
endif::community[]

This setting enables you to manage snapshot content in a more flexible manner compared to using the xref:postgresql-property-snapshot-select-statement-overrides[`snapshot.select.statement.overrides`] property.
```

ifdef::community[]
|[[postgresql-property-snapshot-snapshot-query-mode-custom-name]]<<postgresql-property-snapshot-query-mode-custom-name, `+snapshot.query.mode.custom.name+`>>
|No default

> | When `snapshot.query.mode` is set as `custom`, this setting must be set to specify the name of the custom implementation provided in the `name()` method defined by the 'io.debezium.spi.snapshot.SnapshotQuery' interface.

Reviewer: ```suggestion
| When xref:postgresql-property-snapshot-query-mode[`snapshot.query.mode`] is set as `custom`, use this setting to specify the name of the custom implementation provided in the `name()` method that is defined by the 'io.debezium.spi.snapshot.SnapshotQuery' interface.
```

> See xref:postgresql-custom-snapshot[custom snapshotter SPI].

Reviewer: ```suggestion
For more information, see xref:postgresql-custom-snapshot[custom snapshotter SPI].
```
